src/main/java/hudson/plugins/gradle/enriched/ScanDetailService.java

```diff
@@ -0,0 +1,139 @@
+package hudson.plugins.gradle.enriched;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import hudson.util.Secret;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.util.EntityUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.util.Optional;
+
+public class ScanDetailService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ScanDetailService.class);
+
+    private final static ObjectMapper MAPPER = new ObjectMapper();
+
+    private static final String GRADLE_ENTERPRISE_PUBLIC_SERVER = "https://gradle.com";
+    private static final String URL_CONTEXT_PATH_SCAN_ID = "/s/";
+    private static final String URL_CONTEXT_PATH_API_BUILDS = "/api/builds/";
+
+    private HttpClientFactory httpClientFactory;
+
+    private final boolean isEnrichedSummaryEnabled;
+    private final Secret buildScanAccessToken;
+    private final String buildScanServer;
+    private final int httpClientTimeoutInSeconds;
+    private final int httpClientMaxRetries;
+    private final int httpClientDelayBetweenRetriesInSeconds;
+
+    void setHttpClientFactory(HttpClientFactory httpClientFactory) {
+        this.httpClientFactory = httpClientFactory;
+    }
+
+    public ScanDetailService(EnrichedSummaryConfig config) {
+        this.isEnrichedSummaryEnabled = config.isEnrichedSummaryEnabled();
+        this.buildScanAccessToken = config.getBuildScanAccessKey();
+        this.buildScanServer = config.getBuildScanServer();
+        this.httpClientTimeoutInSeconds = config.getHttpClientTimeoutInSeconds();
+        this.httpClientMaxRetries = config.getHttpClientMaxRetries();
+        this.httpClientDelayBetweenRetriesInSeconds = config.getHttpClientDelayBetweenRetriesInSeconds();
+        this.httpClientFactory = new HttpClientFactory();
+    }
+
+    public Optional<ScanDetail> getScanDetail(String buildScanUrl) {
+        if (isEnrichedSummaryEnabled && buildScanUrl != null) {
+            return Optional.ofNullable(doGetScanDetail(buildScanUrl));
+        }
+
+        return Optional.empty();
+    }
+
+    private ScanDetail doGetScanDetail(String buildScanUrl) {
+        if (buildScanUrl.startsWith(GRADLE_ENTERPRISE_PUBLIC_SERVER)) {
+            // API is not accessible on public server
+            return null;
+        }
+
+        String baseApiUri = getBaseApiUri(buildScanUrl);
+        if (null == baseApiUri || baseApiUri.isEmpty()) {
+            LOGGER.error("Gradle Enterprise API URL can't be resolved");
```

Comment:
I think we don't need logging here. If URL parsing fails in `getBaseApiUri`, then we will already log a warning there. There is no need to call twice. WDYT?

Reply:
Agreed, I added a log message in `getBaseUri` when buildScanId can't be parsed to get fully covered